File: forecasto-client-web/src/api/partner.ts
import apiClient from './client'

export interface PartnerCode {
  id: string
  code: string
  created_at: string
  expires_at: string | null
  used_at: string | null
  used_by_name: string | null
  used_by_email: string | null
  revoked_at: string | null
  recipient_name: string | null
  recipient_email: string | null
  invoiced: boolean
  invoiced_to: string | null
}

export interface PartnerBatch {
  id: string
  name: string
  created_at: string
  expires_at: string | null
  note: string | null
  total_codes: number
  used_codes: number
  available_codes: number
  codes: PartnerCode[]
}

interface PartnerBatchListResponse {
  success: boolean
  batches: PartnerBatch[]
}

interface PartnerBatchResponse {
  success: boolean
  batch: PartnerBatch
}

interface PartnerCodeResponse {
  success: boolean
  code: PartnerCode
}

export const partnerApi = {
  listBatches: async (): Promise<PartnerBatch[]> => {
    const response = await apiClient.get<PartnerBatchListResponse>('/partner/batches')
    return response.data.batches
  },

  getBatch: async (batchId: string): Promise<PartnerBatch> => {
    const response = await apiClient.get<PartnerBatchResponse>(`/partner/batches/${batchId}`)
    return response.data.batch
  },

  updateCodeRecipient: async (batchId: string, codeId: string, recipientName: string | null, recipientEmail: string | null): Promise<PartnerCode> => {
    const response = await apiClient.patch<PartnerCodeResponse>(`/partner/batches/${batchId}/codes/${codeId}/recipient`, {
      recipient_name: recipientName,
      recipient_email: recipientEmail,
    })
    return response.data.code
  },
}

File: forecasto-server/src/forecasto/schemas/partner.py
"""Partner schemas for partner-facing endpoints."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class PartnerCodeResponse(BaseModel):
    """Single code response for partner view."""

    id: str
    code: str
    created_at: datetime
    expires_at: datetime | None = None
    used_at: datetime | None = None
    used_by_name: str | None = None
    used_by_email: str | None = None
    revoked_at: datetime | None = None
    invoiced: bool = False
    invoiced_to: str | None = None


class PartnerBatchResponse(BaseModel):
    """Batch response for partner view with codes and statistics."""

    id: str
    name: str
    created_at: datetime
    expires_at: datetime | None = None
    note: str | None = None
    total_codes: int = 0
    used_codes: int = 0
    available_codes: int = 0
    codes: list[PartnerCodeResponse] = []
